### PR TITLE
StackTrace parser, fix assertion error message

### LINKED ISSUE #4021: factor out the stack parsing nonsense into a reusable library
If we do any more stack parsing, it's time to bite the bullet and just write a library to parse them. It's becoming far too brittle to do it with handcrafted RegExps everywhere.

### LINKED ISSUE #18412: Hide "this could be a framework bug" message when first stack frame is not framework
We should hide the message saying this could be a framework bug if the first stack frame is not in the framework. In fact we should probably figure out which package is to blame and then explicitly name that.

Depends on https://github.com/flutter/flutter/issues/4021.

### PR DESCRIPTION
**Pull request by dnfield**
## Description

Adds a StackTrace parser (fixes #4021) and fixes up the way we print out "this is likely a bug in the framework" message (fixes https://github.com/flutter/flutter/issues/18412).

The main goal here is to change the way we print out requests to file a bug on assertion errors.  It currently _always_ prints that message because Dart has changed the way it throws assertions (there's an additional frame before the one we used to look for).  However, even fixing that still isn't quite right.  The new logic here will only print the message on an `AssertionError` that has at least two frames of framework code before any other user code - the idea being that we either need to add a new assert to the framework, or stop violating our own asserts in such a case.

## Related Issues

#4021
#18412

## Tests

I added the following tests:

Tests for stack parsing
Tests for checking our fault/not our fault on stack traces.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] We do not consider this a breaking change, because although it changes tests, it's only changing the shape of an error message, and so shouldn't nbe considered breaking (/cc @hixie)

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/foundation/stack_frame.dart [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui' show hashValues;
+
+import 'package:meta/meta.dart';
+
+/// A object representation of a frame from a stack trace.
+///
+/// {@tool sample}
```
Comment (dnfield) [inline packages/flutter/lib/src/foundation/stack_frame.dart]: @gspencergoog

Comment (gspencergoog) [reply packages/flutter/lib/src/foundation/stack_frame.dart]: This is correct (for now).  I'll swap it to be "snippet" in my PR after you land.

Comment (gspencergoog) [reply packages/flutter/lib/src/foundation/stack_frame.dart]: ...or you will after I land, depending on who gets there first!  It will cause an analysis error if you forget, so it shouldn't be too hard to find.

File: packages/flutter/test/foundation/error_reporting_test.dart [hunk 2]
```diff
@@ -45,13 +45,19 @@ Future<void> main() async {
 
   final StackTrace sampleStack = await getSampleStack();
 
-  test('Error reporting - pretest', () async {
+  setUp(() async {
```
Comment (dnfield) [inline packages/flutter/test/foundation/error_reporting_test.dart]: Because we need this before any individual test. When we eventually randomize our tests this will be important.

File: packages/flutter/test/foundation/error_reporting_test.dart [hunk 3]
```diff
@@ -45,13 +45,19 @@ Future<void> main() async {
...
     expect(debugPrint, equals(debugPrintThrottled));
     debugPrint = (String message, { int wrapWidth }) {
       console.add(message);
     };
   });
 
+  tearDown(() async {
```
Comment (dnfield) [inline packages/flutter/test/foundation/error_reporting_test.dart]: Ditto.  Without these changes, none of these tests can be run independently.

File: packages/flutter/lib/src/foundation/stack_frame.dart [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+  /// Whether or not this was thrown from a constructor.
+  final bool isConstructor;
+
+  @override
+  int get hashCode => hashValues(number, package, line, column, className, method);
+
+  @override
+  bool operator ==(Object other) {
+    return other is StackFrame &&
```
Comment (Hixie) [inline packages/flutter/lib/src/foundation/stack_frame.dart]: we should compare the runtimeType here; see style guide for a template

File: packages/flutter/lib/src/foundation/stack_frame.dart [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        number == other.number &&
+        package == other.package &&
+        line == other.line &&
+        column == other.column &&
+        className == other.className &&
+        method == other.method;
+  }
+
+  @override
+  String toString() => '$runtimeType{#$number, $packageScheme:$package/$packagePath:$line:$column, className: $className, method: $method}';
```
Comment (Hixie) [inline packages/flutter/lib/src/foundation/stack_frame.dart]: trivial nit: `(` instead of `{` is our convention

**Comment by Hixie** [comment]
LGTM

**Comment by Hixie** [comment]
If you want to overachieve even more, there's other code in the framework that does stack parsing, like `defaultStackFilter`.

File: packages/flutter/lib/src/foundation/stack_frame.dart (hunk 5 above)
Comment (dnfield) [reply packages/flutter/lib/src/foundation/stack_frame.dart]: Done

File: packages/flutter/lib/src/foundation/stack_frame.dart (hunk 4 above)
Comment (dnfield) [reply packages/flutter/lib/src/foundation/stack_frame.dart]: Done.  We should update our style guide - @a14n changed the way this works for removing the avoid_as lint and this + the runtimeType check is actually the convention used now.

File: packages/flutter/lib/src/foundation/stack_frame.dart [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+}
```
Comment (Hixie) [inline packages/flutter/lib/src/foundation/stack_frame.dart]: would be nice to have a way to print a frame in a manner that matches what the compiler does, so that you could take a List<StackFrame>, filter it, then reprint it and it would come out in the same format as the original, just with some frames omitted.

also, check that this works on web and in AOT mode -- maybe the web stack has differently-shaped stack traces, and I know the AOT compiler has different details (e.g. omits line numbers in some cases?).

Comment (dnfield) [reply packages/flutter/lib/src/foundation/stack_frame.dart]: I believe the line number thing was fixed (saw issues for that), but I'll check.

We could also just stuff the whole original string into the object instead of reconstructing it.

Comment (dnfield) [reply packages/flutter/lib/src/foundation/stack_frame.dart]: AOT is missing columns.  Will fix.

File: packages/flutter/lib/src/foundation/stack_frame.dart [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+///
+/// {@tool sample}
+///
+/// For example, a caller that wishes to traverse the stack could use
+///
+/// ```dart
+/// final List<StackFrame> currentFrames = StackFrame.fromStackTrace(StackTrace.current);
+/// ```
+///
+/// To create a traversable parsed stack.
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/stack_frame.dart]: I believe all text that's outside the dart code block will render above the code in dart docs. So this may look awkward there. I would rephrase this to have the prose text all above the code.

File: packages/flutter/lib/src/foundation/stack_frame.dart [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+        assert(method != null),
+        assert(packageScheme != null),
+        assert(package != null),
+        assert(packagePath != null),
+        assert(className != null),
+        assert(isConstructor != null);
+
+  /// A stack frame representing an asynchronous suspension.
+  static const StackFrame asynchronousSuspension = StackFrame(
+    -1,
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/stack_frame.dart]: Should this also be a named parameter? (currently, I'd have to click through to StackFrame to learn what the -1 means.)

**Review by goderbauer — COMMENTED** [review]
NICE!

File: packages/flutter/lib/src/foundation/stack_frame.dart (hunk 7 above)
Comment (dnfield) [reply packages/flutter/lib/src/foundation/stack_frame.dart]: done.

File: packages/flutter/lib/src/foundation/stack_frame.dart (hunk 8 above)
Comment (dnfield) [reply packages/flutter/lib/src/foundation/stack_frame.dart]: I had it as named but it felt awkward.  I could make it named again.

**Comment by dnfield** [comment]
I've updated this for defaultStackFilter.

File: packages/flutter/lib/src/foundation/stack_frame.dart (hunk 4 above)
Comment (Hixie) [reply packages/flutter/lib/src/foundation/stack_frame.dart]: Please don't hesitate to update the style guide.

**Comment by Hixie** [comment]
LGTM
LGTM
LGTM

...silly bot doesn't like my stamps. \*pout\*

File: packages/flutter/lib/src/foundation/stack_frame.dart (hunk 8 above)
Comment (Hixie) [reply packages/flutter/lib/src/foundation/stack_frame.dart]: you could also just use nulls, fwiw.

**Review by Hixie — APPROVED** [review]

**Comment by dnfield** [comment]
I may want to add one more thing to this before landing.  I think we should keep the original source line - it would make resolving https://github.com/flutter/flutter/issues/4174 easier

**Comment by Hixie** [comment]
You can also just do that in a separate PR (e.g. as part of #4174 if you're interested in that one).

**Comment by dnfield** [comment]
Yeah I'll do that in a separate PR.  I'm not sure I'll actually need the source line, but since the parsing is lossy it seems useful to have.  Parsing stack traces shouldn't need to worry too much about an extra copy of a string anyway.

File: packages/flutter/lib/src/foundation/stack_frame.dart [hunk 9]
```diff
@@ -0,0 +1,230 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
Comment (a14n) [inline packages/flutter/lib/src/foundation/stack_frame.dart]: 2020?

File: packages/flutter/lib/src/foundation/stack_frame.dart [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+        .toList();
+  }
+
+  static StackFrame _parseWebFrame(String line) {
+    final bool hasPackage = line.startsWith('package');
+    final RegExp parser = hasPackage
+        ? RegExp(r'^(package:.+) (\d+):(\d+)\s+(.+)$')
+        : RegExp(r'^(.+) (\d+):(\d+)\s+(.+)$');
+    final Match match = parser.firstMatch(line);
+    assert(match != null, 'Expecgted $line to match $parser.');
```
Comment (a14n) [inline packages/flutter/lib/src/foundation/stack_frame.dart]: typo: Expected

File: packages/flutter/lib/src/foundation/stack_frame.dart (hunk 9 above)
Comment (dnfield) [reply packages/flutter/lib/src/foundation/stack_frame.dart]: All files in this repo are dated to 2014 to keep things consistent

File: packages/flutter/lib/src/foundation/stack_frame.dart (hunk 10 above)
Comment (dnfield) [reply packages/flutter/lib/src/foundation/stack_frame.dart]: Saw this right after landing :( will open a PR to fix, thanks for catching that

File: packages/flutter/lib/src/foundation/stack_frame.dart [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+    if (runtimeType != other.runtimeType)
+      return false;
+    return other is StackFrame &&
+        number == other.number &&
+        package == other.package &&
+        line == other.line &&
+        column == other.column &&
+        className == other.className &&
+        method == other.method &&
+        source == other.source;
```
Comment (a14n) [inline packages/flutter/lib/src/foundation/stack_frame.dart]: If I promoted a style it would be to prefer something like:

```dart
    return other is StackFrame
        && other.number == number
        && other.package == package
        && other.line == line
...
```

(starting line with `&& other.`)

It's the pattern mostly used inside `operator==` methods.

Comment (dnfield) [reply packages/flutter/lib/src/foundation/stack_frame.dart]: This is what the style guide actually now says - I just did it wrong here.  I'll update it in the other PR.

File: packages/flutter/test/foundation/assertions_test.dart [hunk 12]
```diff
@@ -352,4 +352,80 @@ void main() {
       expect(summary.value, equals(<String>['Invalid argument(s) (myArgument): Must not be null']));
     }
   });
+
+  test('Identifies user fault', () {
+    // User fault because they called `new Text(null)` from their own code.
+    final StackTrace stack = StackTrace.fromString('''#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:42:39)
```
Comment (a14n) [inline packages/flutter/test/foundation/assertions_test.dart]: The first line could also be on a newline for readability (the new line followin `'''` in multiline string is ignored)

File: packages/flutter/test/foundation/assertions_test.dart [hunk 13]
```diff
@@ -352,4 +352,80 @@ void main() {
...
+    expect(builder.properties[2] is ErrorSpacer, true);
+    final DiagnosticsStackTrace trace = builder.properties[3] as DiagnosticsStackTrace;
+    expect(trace, isNotNull);
+    expect(trace.value, stack);
+  });
+
+  test('Identifies our fault', () {
+    // Our fault because we should either have an assertion in `text_helper.dart`
+    // or we should make sure not to pass bad values into new Text.
+    final StackTrace stack = StackTrace.fromString('''#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:42:39)
```
Comment (a14n) [inline packages/flutter/test/foundation/assertions_test.dart]: ditto
